language/tools/move-cli/tests/testsuite/republish_cyclic_transitive/args.exp

```diff
@@ -0,0 +1,16 @@
+Command `publish src1 -v`:
+Compiling Move modules...
+Found and compiled 4 modules
+Publishing a new module 00000000000000000000000000000042::A
+Publishing a new module 00000000000000000000000000000042::B
+Publishing a new module 00000000000000000000000000000042::C
+Publishing a new module 00000000000000000000000000000042::D
+Command `publish src2 -v`:
+Compiling Move modules...
+Found and compiled 1 modules
+Publishing failed with unexpected error CYCLIC_MODULE_DEPENDENCY
```

Comment:
Yes, I am a bit surprised too. I guess sth is wrong with the `shadow` logic? But I haven't check it.

Reply:
I think I found the issue, and not really sure it is truly fixable. 

The source language does not introspect the bodies of functions for dependency checking (since for perf it doesn't want to look at their function bodies). So the source language doesn't see the circularity
(in this case it would be even harder since it is using the generated "interface" file for 0x42::M)

Reply:
Ah, right, in the interface files, `use 0x42::N` is stripped, which is there in M's source code.

Reply:
I don't think uses are ever considered for dependencies (if they are its just for error message purposes).

Conceptually, uses are just aliases. 

The real issue here is that the interface files (1) don't have function bodies and (2) even if they did they are marked as dep files and the function bodies would be ignored for perf reasons

Reply:
Sorry, I misunderstood `use`. Yes, those are the real issues.

But does this mean that the compiler may produce bytecode that may not pass the bytecode verifier? Well, it might be wrong to say that the bytecode *does not pass* the bytecode verifier because the per-module checkings are all fine, and the linkings are fine too. It is only this cyclic dep that causes some trouble.

Reply:
The compiler will always produce code that might not pass the being able to being published on chain. In other words, the compiler has no on chain view, so it could always produce code that is inconsistent or wrong with what is in a given "chain"

Reply:
Right! I think what I was trying to say is:
1) for module compilation we pass it the module source + dependencies for compilation, and
2) even the dependencies are exactly the same as the ones on-chain,
3) the compiler may output the module bytecode that will be rejected by the bytecode verifier.

I am not sure whether this may happen before republishing is allowed.

Reply:
It sort of could have happened before.

I think it's fine for now and there are ways of dealing with it in the future if it is more of an issue. Its quite the edge case and if you are really concerned about it, you can make all of your files "source files", which might be the common case anyway